islandora-connector-derivative/src/main/java/ca/islandora/alpaca/connector/derivative/DerivativeOptions.java

```diff
@@ -0,0 +1,190 @@
+
+  private static final Logger LOGGER = getLogger(DerivativeOptions.class);
+
+  private static final String DERIVATIVE_LIST_PROPERTY = "derivative.systems.installed";
+  private static final String DERIVATIVE_PREFIX = "derivative";
+  private static final String DERIVATIVE_ENABLED_PROPERTY = "enabled";
+  private static final String DERIVATIVE_INPUT_PROPERTY = "in.stream";
+  private static final String DERIVATIVE_OUTPUT_PROPERTY = "service.url";
+  private static final String DERIVATIVE_CONCURRENT_PROPERTY = "concurrent-consumers";
+  private static final String DERIVATIVE_MAX_CONCURRENT_PROPERTY = "max-concurrent-consumers";
+
+  @Autowired
+  private Environment environment;
+
+  @Value("${" + DERIVATIVE_LIST_PROPERTY + ":#{null}}")
+  private String derivativeSystems;
+
+  /**
+   * Register additional beans for derivative routes.
+   *
+   * @param camelContext
+   *   The camel context to add derivative service routes to.
+   */
+  private void processAllServices(final CamelContext camelContext) {
+    if (derivativeSystems != null && !derivativeSystems.isBlank()) {
+      final var systemNames = derivativeSystems.contains(",") ?
+              Arrays.stream(derivativeSystems.split(",")).filter(o -> !o.isBlank()).map(String::trim)
+                      .collect(Collectors.toSet()) : Set.of(derivativeSystems);
+      for (final var system : systemNames) {
+        final var enabled =
+                environment.getProperty(enabledProperty(system), Boolean.class, false);
+        if (enabled) {
+          startDerivativeService(camelContext, system);
+        } else {
+          LOGGER.debug("Derivative connector (" + system + ") is disabled, skipping");
+        }
+      }
+    }
+  }
+
+  /**
+   * Attempt to start the routes and add them to the camel context.
+   *
+   * @param camelContext
+   *   The current camel context.
+   * @param serviceName
+   *   The derivative service name.
+   */
+  private void startDerivativeService(final CamelContext camelContext, final String serviceName) {
+      final var input = environment.getProperty(inputProperty(serviceName), "");
+      final var output = environment.getProperty(outputProperty(serviceName), "");
+      if (!input.isBlank() && !output.isBlank()) {
+        final int concurrentConsumers = environment.getProperty(concurrentConsumerProperty(serviceName),
+                Integer.class, -1);
+        final int maxConcurrentConsumers = environment.getProperty(maxConcurrentConsumerProperty(serviceName),
+                Integer.class, -1);
+        try {
+          // Add concurrent/max-concurrent
+          final String finalInput = addJmsOptions(addBrokerName(input), concurrentConsumers, maxConcurrentConsumers);
+          // Add connectionClose and other http options.
+          final String finalOutput = addHttpOptions(output);
+          camelContext.addRoutes(new DerivativeConnector(serviceName, finalInput, finalOutput, this));
+        } catch (final Exception e) {
```

Comment:
Yeah, some of these PMD rules are suspect to me and might require some review. 

In this case we have no option as that is the error returned by `camelContext.addRoutes()`

https://github.com/apache/camel/blob/camel-3.5.0/core/camel-api/src/main/java/org/apache/camel/CamelContext.java#L570

Reply:
Do... we want to catch it, or propagate it back up the chain? As in, being configured to start with the derivative connector, but unable to set things up... should possibly propagate back up to kill the app, to be more noticeable instead of just dumping the stacktrace but proceeding after the failure?

... That said, not terribly sure what the call stack would look like, when hitting this... from `setApplicationContext()`, and then into the depths of Spring?

Reply:
I can let it flow back up to `setApplicationContext()` but then it looks like I need to wrap it in a `BeansException` to get it out.
Which I can do but then I'm still catching the `Exception` there and will need to ignore there.

Reply:
Is this only happening because we're not quite matching up with the expectations of the Spring Bean lifecycle? As in: Our `setApplicationContext()` thing here isn't really just setting the context, it's more doing initialization things.

Really, could we not just have the `CamelContext` injected (possibly even in the constructor?), and then do our initialization business with it? ... thinking I might take a shot at this approach (and throw a PR against this one).

Reply:
Go for it, I think I tried that but I could be wrong. I can't quite remember.

Reply:
I've got https://github.com/whikloj/Alpaca/compare/one-app...adam-vessey:fixies

... which seems to work fine for the derivative side of things (that I've tested... really just service image and thumbnail stuff, but not anticipating it blowing up otherwise); however... in trying to test the Fedora stuff, having the exchange go sideways:

```
ERROR 17:53:00.693 [Camel (camel-1) thread #6 - seda://nodeIndex] (DefaultErrorHandler) Failed delivery for (MessageId: queue_islandora-indexing-fcrepo-content_ID_islandora8-46433-1638289432407-3_19_-1_1_1 on ExchangeId: ID-islandora8-1638381174585-0-5). Exhausted after delivery attempt: 6 caught: org.apache.camel.InvalidPayloadException: No body available of type: java.io.InputStream but has value: null of type: null on: JmsMessage[JmsMessageID: ID:islandora8-46433-1638289432407-3:19:-1:1:1]. Caused by: No type converter available to convert from type: null to the required type: java.io.InputStream with value null. Exchange[ID-islandora8-1638381174585-0-5]. Caused by: [org.apache.camel.NoTypeConversionAvailableException - No type converter available to convert from type: null to the required type: java.io.InputStream with value null]. Processed by failure processor: FatalFallbackErrorHandler[Channel[log12]]

Message History (complete message history is disabled)
---------------------------------------------------------------------------------------------------------------------------------------
RouteId              ProcessorId          Processor                                                                        Elapsed (ms)
[FcrepoIndexerNodeI] [FcrepoIndexerNodeI] [from[seda://nodeIndex]                                                        ] [      5270]

Stacktrace
---------------------------------------------------------------------------------------------------------------------------------------

org.apache.camel.InvalidPayloadException: No body available of type: java.io.InputStream but has value: null of type: null on: JmsMessage[JmsMessageID: ID:islandora8-46433-1638289432407-3:19:-1:1:1]. Caused by: No type converter available to convert from type: null to the required type: java.io.InputStream with value null. Exchange[ID-islandora8-1638381174585-0-5]. Caused by: [org.apache.camel.NoTypeConversionAvailableException - No type converter available to convert from type: null to the required type: java.io.InputStream with value null]
	at org.apache.camel.support.MessageSupport.getMandatoryBody(MessageSupport.java:117)
	at org.apache.camel.component.http.HttpProducer.createRequestEntity(HttpProducer.java:576)
	at org.apache.camel.component.http.HttpProducer.createMethod(HttpProducer.java:477)
	at org.apache.camel.component.http.HttpProducer.process(HttpProducer.java:118)
	at org.apache.camel.support.AsyncProcessorConverterHelper$ProcessorToAsyncProcessorBridge.process(AsyncProcessorConverterHelper.java:66)
	at org.apache.camel.processor.SendDynamicProcessor.lambda$process$0(SendDynamicProcessor.java:195)
	at org.apache.camel.impl.engine.DefaultProducerCache.doInAsyncProducer(DefaultProducerCache.java:317)
	at org.apache.camel.processor.SendDynamicProcessor.process(SendDynamicProcessor.java:180)
	at org.apache.camel.processor.errorhandler.RedeliveryErrorHandler$RedeliveryTask.redeliver(RedeliveryErrorHandler.java:788)
	at org.apache.camel.impl.engine.DefaultReactiveExecutor$Worker.schedule(DefaultReactiveExecutor.java:148)
	at org.apache.camel.impl.engine.DefaultReactiveExecutor.scheduleMain(DefaultReactiveExecutor.java:60)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:147)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:287)
	at org.apache.camel.component.seda.SedaConsumer.sendToConsumers(SedaConsumer.java:274)
	at org.apache.camel.component.seda.SedaConsumer.doRun(SedaConsumer.java:187)
	at org.apache.camel.component.seda.SedaConsumer.run(SedaConsumer.java:130)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
        [... and so on ...]
```

Seeing the change from `simple("${null}")` to `constant(null)`, poking around there, but really... unsure. Tests seem to go fine inside of Gradle itself, while building.... not sure if there might be some configuration that I'm missing somewhere or what...

Reply:
Yeah... I think it's something related to the fact that it's being expressed as whatever expression?

```java
        msg.setBody(null);
```
in the `CommonProcessor` seemed to do the trick?; however... then seeing other stuff go sideways.

... as in, looking at fcrepo, seeing things go in like:

```
http://localhost:8080/fcrepo/rest/exchangeProperty(fedoraBaseUrl)/cd/a3/48/f0/cda348f0-5ee5-435f-b0e7-ff3249e10f72
```

... seems like something isn't being templated/replaced correctly? Also in that `CommonProcessor` thing?

Reply:
Oh yeah, weird that it would work in tests but fail in real life. But it looks like I needed to reference the exchange as we came out.
https://github.com/whikloj/Alpaca/commit/2954e4d7d55d5de7489a8d99967f7f9d185e0bc2

Reply:
I added a commit to fix those two things in CommonProcessor, so if you want to try that and then open your PR.

Reply:
So yeah... suspecting that expressions are not expected to be used as they are in the `CommonProcessor` thing... setting explicit values, such as:

```diff
diff --git a/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java b/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
index 80c7b27..33b6063 100644
--- a/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
+++ b/islandora-indexing-fcrepo/src/main/java/ca/islandora/alpaca/indexing/fcrepo/CommonProcessor.java
@@ -17,9 +17,6 @@
  */
 package ca.islandora.alpaca.indexing.fcrepo;

-import static org.apache.camel.language.constant.ConstantLanguage.constant;
-import static org.apache.camel.language.property.ExchangePropertyLanguage.exchangeProperty;
-
 import org.apache.camel.Exchange;
 import org.apache.camel.Processor;

@@ -49,8 +46,8 @@ public class CommonProcessor implements Processor {
         final AS2Event json = msg.getBody(AS2Event.class);
         exchange.setProperty("fedoraBaseUrl", json.getTarget());
         msg.removeHeaders("*", "Authorization");
-        msg.setHeader(config.getFedoraUriHeader(), exchangeProperty("fedoraBaseUrl"));
-        msg.setBody(constant(""));
+        msg.setHeader(config.getFedoraUriHeader(), json.getTarget());
+        msg.setBody(null);
         exchange.setMessage(msg);
     }
 }
```

seems to do the trick. The `msg.setHeader()` and `.setBody()` things are documented as accepting `Object` instances; however, the comparable bits on the builders are described as `Expression` instances... 

... And then it seems like the test is giving a pass with the expressions due to... the tests evaluating the expressions, or something?

Reply:
Threw together https://github.com/whikloj/Alpaca/pull/1 ... seems to work, continuing to generate derivatives and whatnot, and nodes and files being indexed in Fedora; however... there's a number of 404s popping, between the Alpaca and Milliner logs... I'm not terribly familiar with Fedora, so I'm not terribly sure what's expected to be there...